libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/util/CacheResultUtils.kt

```diff
@@ -56,6 +59,153 @@ internal object CacheResultUtils {
         }
     }
 
+    data class CacheResultKeyRouteTraffic<R>(
+        val route: NavigationRoute,
+        val trafficProvider: (RouteLeg) -> List<String>?
+    ) :
+        CacheResultCall<(NavigationRoute, (RouteLeg) -> List<String>?) -> R, R> {
+
+        override fun invoke(f: (NavigationRoute, (RouteLeg) -> List<String>?) -> R) = f(
+            route,
+            trafficProvider
+        )
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (javaClass != other?.javaClass) return false
+
+            other as CacheResultKeyRouteTraffic<*>
+
+            if (route.id != other.route.id) return false
+            if (trafficProvider != other.trafficProvider) return false
+            if (
+                route.directionsRoute.legs()?.size !=
+                other.route.directionsRoute.legs()?.size
+            ) {
+                return false
+            }
+            if (!trafficIsEqual(route, other.route)) return false
+            if (!closuresAreEqual(route, other.route)) return false
+            if (!roadClassesAreEqual(route, other.route)) return false
```

Comment:
How did you choose which parts of the routes to compare?
Here you take into account everything that is somehow affects the presentation on the route line, right?

Reply:
This cache is used by MapboxRouteLineUtils::extractRouteData.  This function extracts traffic and road class data from the route ( see ExtractedRouteData ).  If traffic and/or road class data haven't changed then in the context of this function the route hasn't changed.  Closures were included based on feedback but whether or not they are actually needed is up for debate. 

Since there's less data to compare for route equality (in this context) it should be faster than the normal hashCode and equals methods in DirectionsRoute, which are quite expensive.  This is what the work in this PR is attempting to accomplish.